app/src/common/shared/com/igalia/wolvic/browser/extensions/LocalExtension.java

```diff
@@ -0,0 +1,37 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.igalia.wolvic.browser.extensions;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.igalia.wolvic.browser.Addons;
+import com.igalia.wolvic.utils.SystemUtils;
+
+import mozilla.components.concept.engine.webextension.WebExtensionRuntime;
+
+/**
+ * Feature to install temporary extensions from local add-on files.
+ */
+public class LocalExtension {
+
+    private static final String LOGTAG = SystemUtils.createLogtag(LocalExtension.class);
+
+    /**
+     * Installs the web extension in the runtime through the WebExtensionRuntime install method
+     */
+    public static void install(@NonNull WebExtensionRuntime runtime, @NonNull String extensionId, @NonNull String extensionUrl, Addons addons) {
+        runtime.installWebExtension(extensionId, extensionUrl, webExtension -> {
+            Log.i(LOGTAG, extensionId + " from " + extensionUrl + " Web Extension successfully installed");
+            addons.notifyListeners();
```

Comment:
This made me think, why do we need this `LocalExtension`. Can't we manage everything as "normal" addons?

Reply:
The installation flow for a `LocalExtension` is quite different from a regular addon. The primary difference is that for local extensions, we do not have an "Addon" class instance available prior to installation when we refresh our addons list, as info related to that is populated from the extension manifest only after installation. So the local addon installation needs to happen similar to how `BuiltInExtension`s are installed in the existing code.

## Installation workflow difference:

### For regular addons (flow already exists):

1. [Get list of `Addon`s](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/browser/Addons.kt#L210)
2. User presses the + icon for installation
3. ...UI Confirmations for download and install...
4. [Install method requires passing an existing instance of `Addon`](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/addons/views/AddonsListView.java#L180)
5. [All registered listeners receive `onAddonsUpdated`](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/browser/Addons.kt#L141) through `notifyListeners` in the same file


### For local addons (added through this PR, similar to [BuiltInExtension](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/browser/extensions/BuiltinExtension.java#L26)):

1. [User clicks on downloaded addon](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/ui/views/library/DownloadsView.java#L143)
2. Extension [installed through `WolvicWebExtensionRuntime`](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/browser/extensions/LocalExtension.java#L27), not through AddonManager
3. [All registered listeners receive`onAddonsUpdated`](https://github.com/Igalia/wolvic/blob/e050c25f77eb701e06c511241b035db1a754a8f7/app/src/common/shared/com/igalia/wolvic/browser/extensions/LocalExtension.java#L30) through `notifyListeners`, but this time inside `LocalExtension` class after installation is done

I searched through FirefoxForAndroid codebase and GeckoView API, but could not find a way to construct an `Addon` object from xpi file before installation. A possible change in workflow can be to register to [`installPrompt`](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/WebExtensionController.html#install(java.lang.String):~:text=its%20permissions%20through-,installPrompt) call, but I wasn't sure how to integrate that deeply into GeckoView. So I kept the changes straightforward for now

(Also adding this to the first PR comment)

Reply:
I have also made some minor changes to the comments inside this file to make them more relevant.